File: codes/design/in_memory_file_system.py
import collections
from typing import List


class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isFile = False
        self.content = ""
        self.name = ""

class FileSystem:

    def __init__(self):
        self.root = TrieNode()

    def ls(self, path: str) -> List[str]:
        if path == "/":
            return sorted(self.root.children.keys())

        path_list = path.split("/")
        cur = self.root
        for p in path_list[1:]:
            cur = cur.children[p]

        if cur.isFile:
            return [cur.name]
            
        ret = []
        for dir in cur.children:
            ret.append(dir)
        return sorted(ret)

    def mkdir(self, path: str) -> None:
        cur = self.root
        path_list = path.split('/')
        
        for p in path_list[1:]:
            cur = cur.children[p]
            cur.name = p

    def addContentToFile(self, filePath: str, content: str) -> None:
        cur = self.root
        path_list = filePath.split('/')

        for p in path_list[1:]:
            cur = cur.children[p]
            cur.name = p

        cur.isFile = True
        cur.content += content
        

    def readContentFromFile(self, filePath: str) -> str:
        cur = self.root
        path_list = filePath.split('/')
        for p in path_list[1:]:
            cur = cur.children[p]

        if cur.isFile:
            return cur.content
        return ""
        


# Your FileSystem object will be instantiated and called as such:
# obj = FileSystem()
# param_1 = obj.ls(path)
# obj.mkdir(path)
# obj.addContentToFile(filePath,content)
# param_4 = obj.readContentFromFile(filePath)

File: codes/design/in_memory_file_system.cpp
#include "../base.h"

class FileSystem {
public:
    FileSystem() : root_(new TrieNode()) {
        
    }
    
    vector<string> ls(string path) {
        auto curr = getNode(path);
        
        if (curr->isFile)
            return {split(path, '/').back()};
        
        vector<string> result;
        for (const auto& child : curr->children) {
            result.emplace_back(child.first);
        }
        
        sort(result.begin(), result.end());
        return result;
    }
    
    void mkdir(string path) {
        auto curr = putNode(path);
        curr->isFile = false;
    }
    
    void addContentToFile(string filePath, string content) {
        auto curr = putNode(filePath);
        curr->isFile = true;
        curr->content += content;
    }
    
    string readContentFromFile(string filePath) {
        auto curr = getNode(filePath);
        if (curr)
            return curr->content;
        return "";
    }
    
private:
    struct TrieNode {
        bool isFile;
        unordered_map<string, TrieNode*> children;
        string content;
    };
    
    TrieNode* root_;
    
    vector<string> split(const string& s, const char delim) {
        vector<string> tokens;
        stringstream ss(s);
        string token;
        
        while(getline(ss, token, delim)) {
            if (!token.empty())
                tokens.emplace_back(token);
        }
        
        return tokens;
    }
    
    TrieNode* getNode(const string& path) {
        vector<string> strs = split(path, '/');
        auto curr = root_;
        for (const auto& str : strs) {
            curr = curr->children[str];
        }
        
        return curr;
    }
    
    TrieNode* putNode(const string& path) {
        vector<string> strs = split(path, '/');
        auto curr = root_;
        for (const auto& str: strs) {
            if (!curr->children.count(str))
                curr->children[str] = new TrieNode();
            curr = curr->children[str];
        }
        return curr;
    }
    
};

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem* obj = new FileSystem();
 * vector<string> param_1 = obj->ls(path);
 * obj->mkdir(path);
 * obj->addContentToFile(filePath,content);
 * string param_4 = obj->readContentFromFile(filePath);
 */